activesupport/lib/active_support/local_store.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module LocalStore # :nodoc:
+    module FiberLocal
+      def current
+        Thread.current[:active_support_execution_context] ||= {}
+      end
+    end
+
+    module ThreadLocal
+      def current
+        Thread.current.thread_variable_get(:active_support_execution_context) ||
+          Thread.current.thread_variable_set(:active_support_execution_context, {})
+      end
+    end
+
+    ISOLATION_LEVELS = {
+      fiber: FiberLocal,
+      thread: ThreadLocal,
+    }.freeze
+
+    class << self
+      attr_reader :isolation_level
+
+      def isolation_level=(level)
+        mod = ISOLATION_LEVELS.fetch(level)
+        clear if level != isolation_level && respond_to?(:current, true)
+        singleton_class.define_method(:current, mod.instance_method(:current))
+        singleton_class.send(:private, :current)
+        @isolation_level = level
+      end
+
+      def [](key)
+        current[key]
+      end
+
+      def []=(key, value)
+        current[key] = value
+      end
+
+      def clear
+        current.clear
+      end
+    end
```

Comment:
I don't understand the idea of copying an instance method over, could we swap a class instance variable instead? Provided we make them `def self.current` on the Local modules, could it be something like this?

```suggestion
    class << self
      def isolation_level=(level)
        clear if defined?(@state)
        @state = { fiber: FiberLocal, thread: ThreadLocal }.fetch(level)
      end

      delegate :current, to: :@state
      delegate :[], :[]=, :clear, to: :current
    end
```

Reply:
The reason is because if we do migrate most `Thread.current` to use this, it will likely be a huge hotspot, so I'd like to avoid that extra hash lookup + method call. 

The `define_method` could be simplified though. I could do something like:

```
def current_thread
  ...
end

def current_fiber
  ...
end
```

And then select the active one with `alias_method :current, :current_#{level}`. Would that be clean enough for you?